pkg/cmd/codespace/ssh.go

```diff
@@ -147,6 +171,158 @@ func (a *App) SSH(ctx context.Context, sshArgs []string, opts sshOptions) (err e
+			continue
+		}
+
+		cs := cs
+		fetches += 1
+		go func() {
+			result := sshResult{}
+			defer func() {
+				select {
+				case sshUsers <- result:
+				case <-ctx.Done():
+				}
+			}()
+
+			session, err := a.openSSHSession(ctx, cs, nil)
+			if err != nil {
+				result.err = fmt.Errorf("error connecting to codespace: %w", err)
+				return
+			}
+			defer session.Close()
+
+			//a.StartProgressIndicatorWithLabel(fmt.Sprintf("Fetching SSH Details for %s", cs.Name))
+			_, result.user, err = session.StartSSHServer(ctx)
+			//a.StopProgressIndicator()
+			if err != nil {
+				result.err = fmt.Errorf("error getting ssh server details: %w", err)
+				return
+			}
+
+			result.codespace = cs
+		}()
+	}
+
+	// While the above fetches are running, ensure that the user has keys installed.
+	// That lets us report a more useful error message if they don't.
+	if err = a.ensureAuthorizedKeys(ctx); err != nil {
+		return err
+	}
+
+	ghexec, err := os.Executable()
+	if err != nil {
+		return err
+	}
+
+	t, err := template.New("ssh_config").Parse(heredoc.Doc(`
+		Host cs.{{.Name}}.{{.EscapedRef}}
+			User {{.SSHUser}}
+			ProxyCommand {{.GHExec}} cs ssh -c {{.Name}} --stdio
+			UserKnownHostsFile=/dev/null
+			StrictHostKeyChecking no
+			LogLevel quiet
+			ControlMaster auto
+
+	`))
+	if err != nil {
+		return fmt.Errorf("error formatting template: %w", err)
+	}
+
+	for i := 0; i < fetches; i++ {
+		result := <-sshUsers
+		if result.err != nil {
+			fmt.Fprintf(os.Stderr, "%v\n", result.err)
+			status = cmdutil.SilentError
+			continue
+		}
+
+		conf := codespaceSSHConfig{
+			Name:       result.codespace.Name,
+			EscapedRef: strings.ReplaceAll(result.codespace.GitStatus.Ref, "/", "-"),
+			SSHUser:    result.user,
+			GHExec:     ghexec,
+		}
+		if err := t.Execute(a.io.Out, conf); err != nil {
+			return err
+		}
+	}
+
+	return status
+}
+
+type sshResult struct {
+	codespace *api.Codespace
+	user      string // on success, the remote ssh username; else nil
+	err       error
+}
+
+// codespaceSSHConfig contains values needed to write an OpenSSH host
+// configuration for a single codespace. For example:
+//
+// Host {{Name}}.{{EscapedRef}
+//   User {{SSHUser}
+//   ProxyCommand {{GHExec}} cs ssh -c {{Name}} --stdio
+//
+// EscapedRef is included in the name to help distinguish between codespaces
+// when tab-completing ssh hostnames. '/' characters in EscapedRef are
+// flattened to '-' to prevent problems with tab completion or when the
+// hostname appears in ControlMaster socket paths.
+type codespaceSSHConfig struct {
+	Name       string // the codespace name, passed to `ssh -c`
+	EscapedRef string // the currently checked-out branch
+	SSHUser    string // the remote ssh username
+	GHExec     string // path used for invoking the current `gh` binary
+}
+
+func (a *App) openSSHSession(ctx context.Context, codespace *api.Codespace, liveshareLogger *log.Logger) (*liveshare.Session, error) {
+	if liveshareLogger == nil {
```

Comment:
Move this check into the caller, and establish the precondition that logger != nil. In one of the two callers, it will be dead code, so you can delete it.

At that point, this function will do nothing but call ConnectToLiveShare, so I suggest you inline it entirely. (The augmentation of the error message adds very little, and you may be able to give a more specific augmented error in each caller.)